Lib/test/test__xxsubinterpreters.py

```diff
@@ -1203,6 +1203,58 @@ def test_ids_global(self):
 
         self.assertEqual(cid2, int(cid1) + 1)
 
+    def _assert_interpreters_returned(self, cid, send, recv):
+        actual = set(interpreters.channel_list_interpreters(cid, send=True))
+        self.assertEqual(actual, set(send))
+        actual = set(interpreters.channel_list_interpreters(cid, send=False))
+        self.assertEqual(actual, set(recv))
+
+    def test_channel_list_interpreters_empty(self):
+        # Test for channel with no associated interpreters.
+        cid = interpreters.channel_create()
+        self._assert_interpreters_returned(cid, [], [])
+
+    def test_channel_list_interpreters_mainline(self):
```

Comment:
It may be worth adding a test for the case where an interpreter was associated with a channel end but then got un-associated.  The includes:

* a destroyed interpreter
* an interpreter that called `chan.release()` (where another interpreter is still associated)
   + for one end and for both
* an interpreter that called `chan.close()` (where another interpreter is still associated)
   + for one end and for both

Reply:
Good suggestion, this seems to have highlighted some issues, which effectively stem from the note in the code next to the `_channelends` struct:
"Note that the list entries are never removed for interpreter for which the channel is closed.  This should be a problem in practice.  Also, a channel isn't automatically closed when an interpreter is destroyed."

I've added testcases for each of the cases above which are currently failing. Some questions on each:
- How should we deal with an interpreter being destroyed?
  - Loop over channels and do something like call `_channel_close_interpreter()`? (I don't think this would currently 'just work').
  - Keep track of channels associated with interpreters (i.e. a mapping of interpreters to channelends), so that we can more easily close the channelend interpreter associations?
  - Just handle this in `channel_list_interpreters()` somehow - I tried to find a function I could use to check whether an interpreter ID (int64_t) corresponds to an alive interpreter, the closest I can see is `_PyInterpreterState_LookUpID()` in pystate.c. The problem is that errors are set when the ID is unrecognised...

- What should the behaviour be when listing channel end interpreters after the channel end has been closed/released?
  - If an invalid channel ID is passed, `ChannelNotFoundError` is raised - should this be the behaviour when passing the ID of a closed channel (that is, closed for the calling interpreter, which covers the `channel_release()` case too...)?

Reply:
(Is there a standard practice to indicate the PR should no longer be considered ready to merge? It's currently marked as approved but there are some things here for me to sort out first.)

Reply:
> Is there a standard practice to indicate the PR should no longer be considered ready to merge?

There is a `DO-NOT-MERGE` PR label.  However, I'm not sure you have the ability to modify labels.  Just in case, I've added it. :smile:

Reply:
> Good suggestion, this seems to have highlighted some issues, which effectively stem from the note in the code next to the `_channelends` struct:
> "Note that the list entries are never removed for interpreter for which the channel is closed. This should be a problem in practice. Also, a channel isn't automatically closed when an interpreter is destroyed."

Good to know that comment was helpful. :smile:  Do you think there's anything to change about the situation?

Regardless, would you mind fixing [the text](https://github.com/python/cpython/pull/17323/files#diff-de6e5174f1d465f3b1510ce42234c070R538-R541) while you're here?  I noticed some typos.  It should look like this:

```c
    // Note that the list entries are never removed for any interpreter
    // for which the channel is closed.  This should not be a problem in
    ...
```

I'd like to think that if it "should be a problem in practice" that I would have addressed the problem. :smile:

Reply:
> * How should we deal with an interpreter being destroyed?
>   * Loop over channels and do something like call `_channel_close_interpreter()`? (I don't think this would currently 'just work').
>   * Keep track of channels associated with interpreters (i.e. a mapping of interpreters to channelends), so that we can more easily close the channelend interpreter associations?
>   * Just handle this in `channel_list_interpreters()` somehow - I tried to find a function I could use to check whether an interpreter ID (int64_t) corresponds to an alive interpreter, the closest I can see is `_PyInterpreterState_LookUpID()` in pystate.c. The problem is that errors are set when the ID is unrecognised...

Ah, I think the problem just clicked for me. :smile:  What you're saying is that we currently keep track of channel associations for interpreters that no longer exist.  We need a way to filter those out.  Is that right?

How about just looping over all valid interpreters and call `_channelend_find()` for each one, building the list as you go (inspired by `_channelends_associate()` and `interp_list_all()`):

```c
/* "high"-level channel-related functions */
...

static int
_channel_is_associated(_channels *channels, int64_t cid, int64_t interp, int send)
{
    _PyChannelState *chan = _channels_lookup(channels, cid, NULL);
    ...  # return -1 for error
    _channelend *end = _channelend_find(send ? chan->ends->send : chan->ends->recv, interp, NULL);
    return (end != NULL && end->open);
}

...
/* module level code ********************************************************/
...

static PyObject *
channel_list_interpreters(...)
{
    PyObject *ids, *id;
    PyInterpreterState *interp;

    ids = PyList_New(0);
    ...
    interp = PyInterpreterState_Head();
    while (interp != NULL) {
        id = _PyInterpreterState_GetIDObject(interp);
        ...
        int res = _channel_is_associated(&_globals.channels, cid, id, send);
        if (!res) {
            continue
        }
        if (res < 0) {
            ...
        }
        int res = PyList_Insert(ids, 0, id);
        ...
        interp = PyInterpreterState_Next(interp);
    }
    return ids;
}
```

It probably isn't the most efficient approach, but it's simple, correct, and good enough. :smile:  We can leave optimizing for later.

> * What should the behaviour be when listing channel end interpreters after the channel end has been closed/released?
>   * If an invalid channel ID is passed, `ChannelNotFoundError` is raised - should this be the behaviour when passing the ID of a closed channel (that is, closed for the calling interpreter, which covers the `channel_release()` case too...)?

I'd expect `ChannelClosedError` if the channel is fully closed but still exists.  `ChannelNotFound` should only be for the case that the channel never existed (or has been completely erased from memory).  If the channel is only "released" (still associated with at least one interpreter) then I would expect no error, even if it was released by the current interpreter.  The same goes for the channel ends.  If only one end has interpreters associated then I'd expect to get an empty list for the "empty" end.

Reply:
> Regardless, would you mind fixing [the text](https://github.com/python/cpython/pull/17323/files#diff-de6e5174f1d465f3b1510ce42234c070R538-R541) while you're here? I noticed some typos. It should look like this:
> 
> ```c
>     // Note that the list entries are never removed for any interpreter
>     // for which the channel is closed.  This should not be a problem in
>     ...
> ```
> 
> I'd like to think that if it "should be a problem in practice" that I would have addressed the problem. 😄

I took this comment too literally, that makes sense if it's just the comment that's wrong! I'll fix that up.


> Ah, I think the problem just clicked for me. 😄 What you're saying is that we currently keep track of channel associations for interpreters that no longer exist. We need a way to filter those out. Is that right?

Yes that sounds right. I did have a bit of a go at this without getting very far - I'll take another look with your suggestions to guide me at some point soon.